bittensor/_dataset/dataset_impl.py

```diff
@@ -1,536 +1,490 @@
 
+        """ 
+        Async save of json for storing text hashes
         Args:
-            directory: Map{ Name: str, Hash: str, Size: int }: 
-                The original directory to look up a datafile for.
-
-        Returns:
-            directory: Map{ Name: str, Hash: str, Size: int }: 
-                A random directory that lead to a datafile.
+            path (str):
+                Path of the loaded json
+            include_root (bool):
+                Include self.data_dir as the prefix.
+                    - if True, ths meants shortens the batch and 
+                    specializes it to be with respect to the dataset's 
+                    root path which is in ./bittensor/dataset
+        Returns: 
+            obj (str)
+                Object of the saved JSON.
         """
-        # --- If the size of directory is small, it is leads to data file, return the data file.
-        if file_meta['Size'] <= self.datafile_size_bound:
-            return file_meta
+        
+        if include_root:
+            path = os.path.join(self.data_dir, path)
 
-        # --- Else, the directory leads to more directories, return a random data file within the directories.
-        else:
-            response = self.get_ipfs_directory(self.text_dir, file_meta)
-            # --- Return none if the request failed.
-            if (response == None) or (response.status_code != 200):
-                logger.warning("Failed to retrieve directory, ignoring directory:".ljust(20) + "<blue>{}</blue>".format(file_meta))
-                return None
-            
-            # --- Pick a random sub_directory, run recursion until we have found a data file
-            else:
-                sub_directories = response.json()
-                if sub_directories and 'Links' in sub_directories.keys() and len(sub_directories['Links']) >= 1:
-                    random_sub_directory = random.choice(sub_directories['Links'])
+        # Ensure extension.
+        dir_path = os.path.dirname(path)
+        if path[-len('.json'):] != '.json':
+            path += '.json'
 
-                    # --- Fill the name of the random_sub_directory if it is empty. 
-                    if random_sub_directory['Name'] == '':
-                        random_sub_directory['Name'] = file_meta['Name']
-                        random_sub_directory['Folder'] = file_meta['Folder']
+        # Ensure dictionary.
+        if not os.path.isdir(dir_path):
+            os.makedirs(dir_path)
 
-                    
-                    return self.get_root_text_hash(random_sub_directory)
-                else:
-                    logger.warning("Directory seems empty, ignoring directory:".ljust(20) + "<blue>{}</blue>". format(file_meta))
-        return None
+        # Load default if file does not exist.
+        try:
+            with open(path, 'r') as f:
+                obj = json.load(f)
+        except FileNotFoundError:
```

Comment:
It's possible for this file to be empty, i.e. if someone executes this script and the file is partially created but not written to (as happened to me in my testing). Should add handling for `JSONDecodeError` as well.